File: app/src/main/java/com/futsch1/medtimer/WeekendModePreferencesFragment.java
package com.futsch1.medtimer;

import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.preference.MultiSelectListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceManager;

import com.futsch1.medtimer.helpers.TimeHelper;
import com.futsch1.medtimer.reminders.ReminderProcessor;
import com.takisoft.preferencex.PreferenceFragmentCompat;

import java.util.List;
import java.util.stream.Collectors;

public class WeekendModePreferencesFragment extends PreferenceFragmentCompat {

    @Override
    public void onCreatePreferencesFix(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.weekend_mode_preferences, rootKey);

        setupWeekendMode();
        setupTimePicker();
        setupDays();
    }

    private void setupWeekendMode() {
        Preference preference = getPreferenceScreen().findPreference(PreferencesNames.WEEKEND_MODE);
        if (preference != null) {
            preference.setOnPreferenceChangeListener((preference1, newValue) -> {
                requestReschedule();
                return true;
            });
        }
    }

    private void setupTimePicker() {
        Preference preference = getPreferenceScreen().findPreference(PreferencesNames.WEEKEND_TIME);
        if (preference != null) {
            SharedPreferences defaultSharedPreferences = PreferenceManager.getDefaultSharedPreferences(requireContext());
            preference.setSummary(TimeHelper.minutesToTimeString(requireContext(), defaultSharedPreferences.getInt(PreferencesNames.WEEKEND_TIME, 540)));
            preference.setOnPreferenceClickListener(preference1 -> {
                int weekendTime = defaultSharedPreferences.getInt(PreferencesNames.WEEKEND_TIME, 540);
                new TimeHelper.TimePickerWrapper(getActivity()).show(weekendTime / 60, weekendTime % 60, minutes -> {
                    defaultSharedPreferences.edit().putInt(PreferencesNames.WEEKEND_TIME, minutes).apply();
                    preference1.setSummary(TimeHelper.minutesToTimeString(requireContext(), minutes));
                    requestReschedule();
                });
                return true;
            });
        }
    }

    private void setupDays() {
        Preference preference = getPreferenceScreen().findPreference(PreferencesNames.WEEKEND_DAYS);
        if (preference != null) {
            preference.setSummaryProvider((Preference.SummaryProvider<MultiSelectListPreference>) preference1 -> {
                @SuppressWarnings("java:S6204") // Using SDK 33
                List<CharSequence> values = preference1.getValues().stream().map(s -> preference1.getEntries()[Integer.parseInt(s) - 1]).collect(Collectors.toList());
                return String.join(", ", values);
            });
            preference.setOnPreferenceChangeListener((preference12, newValue) -> {
                requestReschedule();
                return true;
            });
        }
    }

    private void requestReschedule() {
        ReminderProcessor.requestReschedule(requireContext());
    }
}

File: app/src/main/java/com/futsch1/medtimer/preferences/WeekendModePreferencesFragment.kt
package com.futsch1.medtimer.preferences

import android.os.Bundle
import android.view.View
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.preference.MultiSelectListPreference
import androidx.preference.Preference
import androidx.preference.Preference.SummaryProvider
import androidx.preference.PreferenceFragmentCompat
import com.futsch1.medtimer.R
import com.futsch1.medtimer.helpers.TimeHelper.TimePickerWrapper
import com.futsch1.medtimer.helpers.TimeHelper.minutesToTimeString
import com.futsch1.medtimer.model.UserPreferences
import com.futsch1.medtimer.preferences.PreferencesDataSource.Companion.WEEKEND_DAYS
import com.futsch1.medtimer.preferences.PreferencesDataSource.Companion.WEEKEND_MODE
import com.futsch1.medtimer.preferences.PreferencesDataSource.Companion.WEEKEND_TIME
import com.futsch1.medtimer.reminders.ReminderContext
import com.futsch1.medtimer.reminders.ReminderProcessorBroadcastReceiver.Companion.requestScheduleNextNotification
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import java.time.LocalTime
import java.util.stream.Collectors
import javax.inject.Inject

@AndroidEntryPoint
class WeekendModePreferencesFragment : PreferenceFragmentCompat() {
    @Inject
    lateinit var preferencesDataSource: PreferencesDataSource

    private var currentSettings: UserPreferences? = null

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        preferenceManager.preferenceDataStore = preferencesDataSource

        setPreferencesFromResource(R.xml.weekend_mode_preferences, rootKey)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupWeekendMode()
        setupTimePicker()
        setupDays()

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                preferencesDataSource.preferences.collect { settings ->
                    currentSettings = settings
                    updateTimePicker(settings)
                }
            }
        }
    }

    private fun setupWeekendMode() {
        val preference = preferenceScreen.findPreference<Preference?>(WEEKEND_MODE)
        preference?.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, _ ->
            requestReschedule()
            true
        }
    }

    private fun updateTimePicker(settings: UserPreferences) {
        val preference = preferenceScreen.findPreference<Preference?>(WEEKEND_TIME)
        preference?.setSummary(minutesToTimeString(requireContext(), settings.weekendTime.hour * 60 + settings.weekendTime.minute.toLong()))
    }

    private fun setupTimePicker() {
        val preference = preferenceScreen.findPreference<Preference?>(WEEKEND_TIME)
        if (preference != null) {
            preference.onPreferenceClickListener = Preference.OnPreferenceClickListener { preference1: Preference? ->
                val weekendTime = currentSettings?.weekendTime ?: LocalTime.of(9, 0)
                TimePickerWrapper(requireActivity()).show(weekendTime.hour, weekendTime.minute) { minutes: Int ->
                    preferencesDataSource.setWeekendTime(LocalTime.of(minutes / 60, minutes % 60))
                    preference1!!.setSummary(minutesToTimeString(requireContext(), minutes.toLong()))
                    requestReschedule()
                }
                true
            }
        }
    }

    private fun setupDays() {
        val preference = preferenceScreen.findPreference<Preference?>(WEEKEND_DAYS)
        if (preference != null) {
            preference.setSummaryProvider(SummaryProvider { preference1: MultiSelectListPreference? ->
                preference1!!.values.stream().map { s: String? -> preference1.entries[s!!.toInt() - 1] }
                    .collect(Collectors.toList()).joinToString(", ")

            })
            preference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, _ ->
                requestReschedule()
                true
            }
        }
    }

    private fun requestReschedule() {
        val context = getContext()
        if (context != null) {
            requestScheduleNextNotification(ReminderContext(context))
        }
    }
}
